### PR TITLE
WIP New approach

### PR DESCRIPTION
**Pull request by MSusik**

### CODE REVIEW

File: jsonalchemy/wrappers.py [hunk 1]
```diff
@@ -34,98 +35,150 @@ def validate(data, schema, *args, **kwargs):
     types = {
         'object': (dict, JSONDict),
         'array': (list, JSONList),
-        'string': (str, unicode),
+        'string': (str, JSONString),
```
Comment (jirikuncar) [inline jsonalchemy/wrappers.py]: I would keep the unicode there too.

File: jsonalchemy/wrappers.py [hunk 2]
```diff
@@ -34,98 +35,150 @@ def validate(data, schema, *args, **kwargs):
...
-    def __getitem__(self, name):
-        return self.dict[name]
+    def validate(self):
+        return validate(self, self.schema)
 
-    def __setitem__(self, name, value):
-        has_key = name in self.dict
-        old_value = self.dict.get(name, None)
 
+class JSONDict(dict, JSONBase):
```
Comment (jirikuncar) [inline jsonalchemy/wrappers.py]: can you use `collections.Mapping` instead of `dict`?

File: jsonalchemy/wrappers.py [hunk 3]
```diff
@@ -34,98 +35,150 @@ def validate(data, schema, *args, **kwargs):
...
-        try:
-            validate(self, self.schema)
-        except ValidationError as e:
-            # rollback the dict modification
-            if has_key:
-                self.dict[name] = old_value
-            else:
-                del self.dict[name]
-            raise e
+class JSONList(list, JSONBase):
```
Comment (jirikuncar) [inline jsonalchemy/wrappers.py]: the same about `collections.Iterable`

**Comment by jirikuncar** [comment]
I'm not really sure about wrapping immutable types.

## 

cc @jalavik @jacquerie

File: jsonalchemy/wrappers.py (hunk 2 above)
Comment (MSusik) [inline jsonalchemy/wrappers.py]: That's the point: This implementation inherits from `dict` in order to have all the methods already implemented and a proper `update` that I can use in `__exit__`.

File: jsonalchemy/wrappers.py (hunk 3 above)
Comment (MSusik) [inline jsonalchemy/wrappers.py]: Same as above.

File: jsonalchemy/wrappers.py [hunk 4]
```diff
@@ -34,98 +35,150 @@ def validate(data, schema, *args, **kwargs):
...
+    def validate(self):
+        return validate(self, self.schema)
 
-    def __setitem__(self, name, value):
-        has_key = name in self.dict
-        old_value = self.dict.get(name, None)
 
+class JSONDict(dict, JSONBase):
+
+    def __new__(self, mapping=None, schema=None):
```
Comment (jirikuncar) [inline jsonalchemy/wrappers.py]: `__new__` takes `cls` (class) as first argument and not `self` (https://docs.python.org/2/reference/datamodel.html#object.__new__).

**Comment by jacquerie** [comment]
One use case I see is defining a DOI schema, which is a `JSONString` with some `format` validation, or an international telephone number.

**Comment by MSusik** [comment]
Well, there are more use cases. ORCID validation, email validation, url validation, etc.

**Comment by jacquerie** [comment]
A little preliminary testing on the performance of wrapping immutables, using [this code](https://github.com/jacquerie/jsonalchemy/tree/6de3347db6ac26c9b513b3dd03113063ff176596):

``` python
In [1]: from jsonalchemy import performance

In [2]: %timeit performance.wrap_perf
10000000 loops, best of 3: 58.7 ns per loop

In [3]: %timeit performance.no_wrap_perf
The slowest run took 19.18 times longer than the fastest. This could mean that an intermediate result is being cached
10000000 loops, best of 3: 62.2 ns per loop

In [4]: %timeit performance.no_wrap_perf
10000000 loops, best of 3: 62 ns per loop

In [5]: %timeit performance.wrap_perf
10000000 loops, best of 3: 58.8 ns per loop
```

A little higher, but less than a 10% increase. Now I will test memory use.

**Comment by jacquerie** [comment]
When I measure memory usage I get some weird results:

```
Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    45     13.0 MiB      0.0 MiB   @profile
    46                             def wrap_perf():
    47     13.0 MiB      0.0 MiB       schema = load_schema_from_url(
    48     13.0 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    49
    50    381.8 MiB    368.8 MiB       data = JSONArray([str(el) for el in xrange(1000000)], schema)


Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    52     17.6 MiB      0.0 MiB   @profile
    53                             def no_wrap_perf():
    54     17.6 MiB      0.0 MiB       schema = load_schema_from_url(
    55     17.6 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    56
    57    382.2 MiB    364.6 MiB       data = JSONArrayNoWrapping([str(el) for el in xrange(1000000)], schema)

Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    52     17.9 MiB      0.0 MiB   @profile
    53                             def no_wrap_perf():
    54     17.9 MiB      0.0 MiB       schema = load_schema_from_url(
    55     17.9 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    56
    57    389.8 MiB    371.9 MiB       data = JSONArrayNoWrapping([str(el) for el in xrange(1000000)], schema)


Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    45     18.2 MiB      0.0 MiB   @profile
    46                             def wrap_perf():
    47     18.2 MiB      0.0 MiB       schema = load_schema_from_url(
    48     18.2 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    49
    50    382.8 MiB    364.7 MiB       data = JSONArray([str(el) for el in xrange(1000000)], schema)
```

In one case the wrapped version consumes more memory, in the other the **unwrapped** version is consuming more. I _think_ that this means that the amount of memory consumed by immutable type wrappers is negligible.

**Comment by MSusik** [comment]
cc @jirikuncar  :arrow_up: 

As yesterday there was an assumption without any proof that wrapping will cause big overhead.

**Comment by egabancho** [comment]
To me wrapping inmutables is not the big issue, instead the small mess inside `__new__` and `__init__` is something we should look closer and find a nicer/safer solution (if we decided to continue with this approach), maybe using metaclasses instead of 'raping' `__new__`?

**Comment by tiborsimko** [comment]
> In [2]: %timeit performance.wrap_perf
> 10000000 loops, best of 3: 58.7 ns per loop
> 
> In [3]: %timeit performance.no_wrap_perf
> The slowest run took 19.18 times longer than the fastest. This could mean that an intermediate result is being cached
> 10000000 loops, best of 3: 62.2 ns per loop

An interesting observation: `wrap_perf` is faster than `no_wrap_perf`? Shouldn't _no_ wrapping be always faster than _some_ wrapping?  Are the tests OK?

**Comment by tiborsimko** [comment]
> The slowest run took 19.18 times longer than the fastest. This could mean that an intermediate result is being cached

19x longer seems excessive...  Do you use some caching or memoising?  If not, you may be perhaps cons'ing too much (seeing large `xrange(...)` and list additions) and may be bitten by [this Python GC problem](http://tiborsimko.org/python-garbage-collection-issue.html)?  You can try to store results in a variable and disable or force GC...

**Comment by MSusik** [comment]
> To me wrapping inmutables is not the big issue, instead the small mess inside **new** and **init** is something we should look closer and find a nicer/safer solution (if we decided to continue with this approach), maybe using metaclasses instead of 'raping' **new**?

I agree, it looks horrible. However, I can't think of another way of subclassing from `__dict__` directly, extending it with a schema and keeping nice class structure. I'd love to hear some suggestion for improvements.

**Comment by jacquerie** [comment]
> An interesting observation: wrap_perf is faster than no_wrap_perf? Shouldn't no wrapping be always faster than some wrapping? Are the tests OK?

Uh! I didn't notice it was the other way around from what it's reasonable. _Maybe_ it's because redefining `append` on `JSONArrayNoWrapping` requires a method search every time the method is needed?

> 19x longer seems excessive... Do you use some caching or memoising? If not, you may be perhaps cons'ing too much (seeing large xrange(...) and list additions) and may be bitten by this Python GC problem? You can try to store results in a variable and disable or force GC...

I only had that message on the first run I made. I agree that the constructor is doing too much consing, in fact it should probably be rewritten as a list comprehension and retested.

**Comment by jacquerie** [comment]
Surprising as it is, I think I confirm the previous result as far as memory usage goes. With [this code](https://github.com/jacquerie/jsonalchemy/blob/d9289051e539abb7d43300bc5002c28db9e943ac/jsonalchemy/performance.py) I get:

``` python
Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    46     12.9 MiB      0.0 MiB   @profile
    47                             def wrap_perf():
    48     12.9 MiB      0.0 MiB       schema = load_schema_from_url(
    49     12.9 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    50
    51    381.8 MiB    368.9 MiB       data = JSONArray([str(el) for el in xrange(1000000)], schema)

Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    54     17.5 MiB      0.0 MiB   @profile
    55                             def no_wrap_perf():
    56     17.5 MiB      0.0 MiB       schema = load_schema_from_url(
    57     17.5 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    58
    59    382.2 MiB    364.6 MiB       data = JSONArrayNoWrapping([str(el) for el in xrange(1000000)], schema)


Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    54     17.9 MiB      0.0 MiB   @profile
    55                             def no_wrap_perf():
    56     17.9 MiB      0.0 MiB       schema = load_schema_from_url(
    57     17.9 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    58
    59    382.4 MiB    364.5 MiB       data = JSONArrayNoWrapping([str(el) for el in xrange(1000000)], schema)


Filename: jsonalchemy/performance.py

Line #    Mem usage    Increment   Line Contents
================================================
    46     18.2 MiB      0.0 MiB   @profile
    47                             def wrap_perf():
    48     18.2 MiB      0.0 MiB       schema = load_schema_from_url(
    49     18.2 MiB      0.0 MiB           abs_path('schemas/performance.json'))
    50
    51    382.7 MiB    364.5 MiB       data = JSONArray([str(el) for el in xrange(1000000)], schema)
```

(to replicate: `$ python jsonalchemy/performance.py`.)

**Comment by jirikuncar** [comment]
New direction has been chosen. Thanks for the prototype.

**Pull request closed without merging**